docs/architecture-decision-records/20220121-conduit-plugin-architecture.md

```diff
@@ -0,0 +1,321 @@
+# Conduit Plugin Architecture
+
+## Summary
+
+This document describes the design for the Conduit plugin architecture. It spans from how Conduit communicates with built-in
+and standalone plugins, where and how the interface is defined, how the plugin SDK is defined.
+
+## Context
+
+We want to provide a good development experience to plugin developers, we also want to be able to ship Conduit with real
+built-in plugins (compiled into the Conduit binary), we want to make it as easy as possible to write plugins in any
+programming language and we want the SDK to be decoupled from Conduit and be able to change without changing Conduit
+itself.
+
+## Decision
+
+The decision can be broken up into 4 parts:
+
+- We defined the plugin interface in gRPC
+- We defined the interface used internally in Conduit to communicate with plugins
+- We introduced plugin registries to support built-in and standalone plugins
+- We defined the interface for the plugin SDK
+
+### Conduit plugin interface (gRPC)
+
+The Conduit plugin interface is defined in gRPC, is standalone (does not depend on Conduit or the SDK) and lives
+in https://github.com/conduitio/conduit-plugin. This repository is the only common thing between Conduit and a plugin,
+meaning that Conduit uses the interface to interact with plugins and plugins implement the interface.
+
+The proto files define the messages and gRPC interface that needs to be implemented by the connector plugin. The
+interface is uploaded to the [buf schema registry](https://buf.build/conduitio/conduit-plugin), which allows us to
```

Comment:
I assume your concern is that the commercial service gets discontinued, right? Given that the product is being adopted by the community (3.5k stars on GitHub and [rising](https://star-history.com/#bufbuild/buf&Date)) and given that it solves a problem, I feel like the benefits outweigh the risks. Buf massively reduces the pain when working with proto files, to list a few benefits: it lints our proto schema, simplifies code generation (especially with remote generation), takes care of external dependencies, since it's a Go tool we can easily ensure everyone is using the same buf version, it provides GitHub actions, it hosts user-readable documentation. All in all, it's much easier to use than vanilla protoc.

> It also looks like they only support Go client generation right now

That's true for remote generation, but hosting the schema in the schema registry allows you to generate code in the language of your choice with ease (see the readme in [conduit-plugin](https://github.com/ConduitIO/conduit-plugin#generating-code-for-other-languages)). We could easily create a repository with a `buf.gen.yaml` file and a scheduled GitHub action that generates the code in any other language (although I secretly hope that buf will provide remote generation for other languages in the future).

I suggest sticking with buf and dealing with it when/if it becomes a problem. Let me know if you still feel like this is a risk we want to avoid right now.

Reply:
Ok, that's fine with me 👍